### PR TITLE
[PART 2] Job to push priority cases to judges

### PR DESCRIPTION
**Pull request by hschallhorn**
bumps #14604

Part 1: #14691

### Description
Job that pushes priority cases to a judge rather than waiting for them to request cases. This will distribute cases to all judges whose teams that have `accepts_priority_pushed_cases` enabled. The first step distributes all priority cases tied to a judge without limit. The second step distributes remaining general population cases (cases not tied to an active judge) while attempting to even out the number of priority cases all judges have received over one month. "Fairness" is shown below

Judge|Distributions this month|Cases to distribute|After distribution
---|---|---|---
A|50|0|50
B|28|0|28
C|22|2|24
D|0|24|24
E|0|24|24

### Acceptance Criteria
- [ ] A check occurs to ensure VLJ is authorized to receive auto-distributed cases
- [ ] VLJs that are authorized to receive auto-distributed cases receive priority cases tied to them as well as priority gen-pop cases 
   - [ ] Gen-pop priority cases are fairly distributed among judges to aim for equal monthly distributions
   - [ ] VLJs receive priority cases tied to them without limit

### Testing Plan
1. TODO

### Code Documentation Updates
- [x] Add or update code comments at the top of the class, module, and/or component.

### CODE REVIEW

File: app/jobs/push_priority_appeals_to_judges_job.rb [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  # even case counts over a full month (or as close as we can get to it)
+  def priority_target
+    @priority_target ||= begin
+      distribution_counts = eligible_judge_priority_distributions_this_month.values
+      target = (distribution_counts.sum + ready_priority_appeals_count) / distribution_counts.count
+
+      while distribution_counts.any? { |distribution_count| distribution_count > target }
+        distribution_counts = distribution_counts.reject { |distribution_count| distribution_count > target }
+        target = (distribution_counts.sum + ready_priority_appeals_count) / distribution_counts.count
+      end
```
Comment (lomky) [inline app/jobs/push_priority_appeals_to_judges_job.rb:71]: This is kinda hard to parse - could we add an explainer for this loop? the combination of changing both the discribution_counts and target is throwing me, I think

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb [hunk 2]
```diff
@@ -0,0 +1,824 @@
...
+        priority_push: false,
+        completed_at: 1.day.ago
+      ).tap { |distribution| distribution.update!(status: :completed) }
+    end
+    let!(:pending_priority_distribution) do
+      create(
+        :distribution,
+        judge: create(:user, :with_vacols_judge_record),
+        priority_push: true,
+        completed_at: 1.day.ago
```
Comment (lomky) [inline spec/jobs/push_priority_appeals_to_judges_job_spec.rb:797]: if its pending, should it have completed_at?

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb [hunk 3]
```diff
@@ -0,0 +1,824 @@
...
+
+    it "only returns judges of active judge teams" do
+      expect(subject).to match_array([judge_with_team])
+    end
+  end
+
+  context ".judge_priority_distributions_this_month" do
+    let(:batch_size) { 20 }
+    let!(:judge_with_no_priority_distributions) do
+      create(:user, :with_vacols_judge_record).tap do |judge|
```
Comment (lomky) [inline spec/jobs/push_priority_appeals_to_judges_job_spec.rb:710]: why does this one `tap` but the next ones dont? :eyes:

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb (hunk 2 above)
Comment (lomky) [reply spec/jobs/push_priority_appeals_to_judges_job_spec.rb:797]: same for line 737

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb [hunk 4]
```diff
@@ -0,0 +1,824 @@
...
+    let!(:judge_with_nonready_priority_cases) do
+      create(:user).tap do |judge|
+        create(:staff, :judge_role, user: judge)
+        vacols_case = create(
+          :case,
+          :aod,
+          bfd19: 1.year.ago,
+          bfac: "3",
+          bfmpro: "ACT",
+          bfcurloc: "not ready",
```
Comment (lomky) [inline spec/jobs/push_priority_appeals_to_judges_job_spec.rb:99]: :laughing:

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb [hunk 5]
```diff
@@ -0,0 +1,824 @@
...
+
+    it "should only distribute the ready priority cases tied to a judge" do
+      expect(subject.count).to eq eligible_judges.count
+      expect(subject.map { |dist| dist.statistics["batch_size"] }).to match_array [2, 0, 0]
+      distributed_cases = DistributedCase.where(distribution: subject)
+      expect(distributed_cases.count).to eq 2
+      expect(distributed_cases.map(&:case_id)).to match_array [ready_priority_bfkey, ready_priority_uuid]
+      expect(distributed_cases.map(&:docket)).to match_array ["legacy", Constants.AMA_DOCKETS.hearing]
+      expect(distributed_cases.map(&:priority).uniq).to match_array [true]
+      expect(distributed_cases.map(&:genpop).uniq).to match_array [false]
```
Comment (lomky) [inline spec/jobs/push_priority_appeals_to_judges_job_spec.rb:140]: I would appreciate some additional comments inside of each `it` block, breaking down what the expects are interrogating? 

like these are pretty straightforward:
```ruby
      expect(subject.count).to eq eligible_judges.count
      expect(subject.map { |dist| dist.statistics["batch_size"] }).to match_array [2, 0, 0]
```
but I don't intuitively understand what we are asserting about the distributed cases here
```
      distributed_cases = DistributedCase.where(distribution: subject)
      expect(distributed_cases.count).to eq 2
      expect(distributed_cases.map(&:case_id)).to match_array [ready_priority_bfkey, ready_priority_uuid]
      expect(distributed_cases.map(&:docket)).to match_array ["legacy", Constants.AMA_DOCKETS.hearing]
      expect(distributed_cases.map(&:priority).uniq).to match_array [true]
      expect(distributed_cases.map(&:genpop).uniq).to match_array [false]
```

(same for the other `it` blocks!)

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb [hunk 6]
```diff
@@ -0,0 +1,824 @@
...
+          :disposition_held,
+          folder_nr: vacols_case.bfkey,
+          hearing_date: 5.days.ago.to_date
+        )
+        vacols_case
+      end
+    end
+    let!(:ready_priority_hearing_cases) do
+      (1..5).map do |i|
+        appeal = FactoryBot.create(:appeal,
```
Comment (lomky) [inline spec/jobs/push_priority_appeals_to_judges_job_spec.rb:186]:
```suggestion
        appeal = create(:appeal,
```

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb [hunk 7]
```diff
@@ -0,0 +1,824 @@
...
+
+    let(:priority_count) { Appeal.count + VACOLS::Case.count }
+    let(:priority_target) { (priority_count + judge_distributions_this_month.sum) / judges.count }
+
+    before do
+      allow_any_instance_of(PushPriorityAppealsToJudgesJob)
+        .to receive(:eligible_judge_priority_distributions_this_month).and_return(
+          judges.each_with_index.map { |judge, i| [judge.id, judge_distributions_this_month[i]] }.to_h
+        )
+    end
```
Comment (lomky) [inline spec/jobs/push_priority_appeals_to_judges_job_spec.rb:225]: Combine with the `before` on line 145?

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb [hunk 8]
```diff
@@ -0,0 +1,824 @@
...
+      expect(distributed_cases.map(&:genpop).uniq.compact).to match_array [true]
+      expect(distributed_cases.pluck(:docket).uniq).to match_array(Constants::AMA_DOCKETS.keys.unshift("legacy"))
+      expect(distributed_cases.group(:docket).count.values.uniq).to match_array [5]
+
+      judges.each_with_index do |judge, i|
+        target_distributions = priority_target - judge_distributions_this_month[i]
+        distribution = subject.detect { |dist| dist.judge_id == judge.id }
+        expect(distribution.statistics["batch_size"]).to eq target_distributions
+        distributed_cases = DistributedCase.where(distribution: distribution)
+        expect(distributed_cases.count).to eq target_distributions
```
Comment (lomky) [inline spec/jobs/push_priority_appeals_to_judges_job_spec.rb:241]: Can we pull this out into a distinct `it` statement, since it reinvokes `subject` anyway?

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb [hunk 9]
```diff
@@ -0,0 +1,824 @@
...
+      let(:distribution_counts) { Array.new(eligible_judge_count, number_of_distributions) }
+
+      context "when there have been no distributions this month" do
+        let(:number_of_distributions) { 0 }
+
+        it_behaves_like "even distributions this month"
+      end
+
+      context "when there have been some distributions this month" do
+        let(:number_of_distributions) { 10 }
```
Comment (lomky) [inline spec/jobs/push_priority_appeals_to_judges_job_spec.rb:300]: Suggestion: this is some multiple of the number of judges, to make the distributions even, correct? If so, can we change it to be based on the judge count variable instead of hardcoded?

Comment (lomky) [reply spec/jobs/push_priority_appeals_to_judges_job_spec.rb:300]: (same for all of the determining numbers in this context)

Comment (lomky) [reply spec/jobs/push_priority_appeals_to_judges_job_spec.rb:300]: Arguing with myself - maybe this is better? I'm trying to think what we would want this to look like if we ever had to tweak this job at future board requests, and how hard updating this would be

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb [hunk 10]
```diff
@@ -0,0 +1,824 @@
...
+
+      it "evens out over multiple calls" do
+        4.times do
+          # Mock distributing cases each week
+          target_distributions = PushPriorityAppealsToJudgesJob.new.eligible_judge_target_distributions_with_leftovers
+          @distribution_counts.merge!(target_distributions) { |_, prev_dists, target_dist| prev_dists + target_dist }
+        end
+        final_counts = @distribution_counts.values.uniq
+        expect(final_counts.max - final_counts.min).to be <= 1
+        expect(final_counts.count).to be <= 2
```
Comment (lomky) [inline spec/jobs/push_priority_appeals_to_judges_job_spec.rb:371]:
```suggestion
        # Expect no more than two distinct counts, no more than 1 apart
        expect(final_counts.max - final_counts.min).to be <= 1
        expect(final_counts.count).to be <= 2
```

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb [hunk 11]
```diff
@@ -0,0 +1,824 @@
...
+        expect(subject).to eq 0
+      end
+    end
+
+    context "when the number of cases are not evenly distributable bewteen all judges" do
+      let(:leftover_cases_count) { target_distributions.count - 1 }
+      let(:priority_count) { target_distributions.values.sum + leftover_cases_count }
+      let(:target_distributions) { to_judge_hash([5, 10, 15, 20, 25]) }
+
+      it "returns no leftover cases" do
```
Comment (lomky) [inline spec/jobs/push_priority_appeals_to_judges_job_spec.rb:413]:
```suggestion
      it "returns the correct count of leftover cases" do
```

File: app/jobs/push_priority_appeals_to_judges_job.rb (hunk 1 above)
Comment (hschallhorn) [reply app/jobs/push_priority_appeals_to_judges_job.rb:71]:
```ruby
      # If there are any judges that have previous distributions that are MORE than the currently calculated priority
      # target, no target will be large enough to get all other judges up to their number of cases. Remove them from
      # consideration and recalculate the target for all other judges.
```
Sound good?

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb (hunk 5 above)
Comment (hschallhorn) [reply spec/jobs/push_priority_appeals_to_judges_job_spec.rb:140]: I'm unsure what to put here. I found myself putting comments that said "Ensure the distributed cases were all priority cases", but that feels unhelpful. Is there better wording you'd suggest? Or other information that I'm not thinking of?

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb (hunk 3 above)
Comment (hschallhorn) [reply spec/jobs/push_priority_appeals_to_judges_job_spec.rb:710]: 👀 what on earth? `create() do ||` works?! Removing the tap then I guess?

File: app/jobs/push_priority_appeals_to_judges_job.rb (hunk 1 above)
Comment (lomky) [reply app/jobs/push_priority_appeals_to_judges_job.rb:77]: Perfect!

File: spec/jobs/push_priority_appeals_to_judges_job_spec.rb (hunk 5 above)
Comment (lomky) [reply spec/jobs/push_priority_appeals_to_judges_job_spec.rb]: In particular, I would comment something similar to what I would write in an `it` statement that just encapsulated them - like:
